tt_metal/impl/dispatch/kernels/packet_queue.hpp

```diff
@@ -136,191 +140,223 @@ class packet_queue_state_t {
-        this->remote_rptr_sent_update_addr =
-            STREAM_REG_ADDR(NUM_PTR_REGS_PER_OUTPUT_QUEUE*adjusted_remote_queue_id,
-                            STREAM_REMOTE_DEST_BUF_SPACE_AVAILABLE_UPDATE_REG_INDEX);
-        this->remote_rptr_cleared_update_addr =
-            STREAM_REG_ADDR(NUM_PTR_REGS_PER_OUTPUT_QUEUE*adjusted_remote_queue_id+1,
-                            STREAM_REMOTE_DEST_BUF_SPACE_AVAILABLE_UPDATE_REG_INDEX);
-
-        this->remote_ready_status_addr = STREAM_REG_ADDR(remote_queue_id, STREAM_REMOTE_SRC_REG_INDEX);
-        this->local_ready_status_ptr = reinterpret_cast<volatile uint32_t*>(
-            STREAM_REG_ADDR(queue_id, STREAM_REMOTE_SRC_REG_INDEX));
     }
 
     inline uint8_t get_queue_id() const {
         return this->queue_id;
     }
 
     inline uint32_t get_queue_local_wptr() const {
-        return *this->local_wptr_val;
-    }
-
-    inline void advance_queue_local_wptr(uint32_t num_words) {
-        if (this->queue_is_input) {
-            *this->local_wptr_update = num_words << REMOTE_DEST_BUF_WORDS_FREE_INC;
-        } else {
-            this->local_wptr += num_words;
-        }
+        return *this->wptr;
     }
 
     inline uint32_t get_queue_local_rptr_sent() const {
-        return *this->local_rptr_sent_val;
+        return *this->rptr_sent;
     }
 
     inline uint32_t get_queue_local_rptr_cleared() const {
-        return *this->local_rptr_cleared_val;
+        return *this->rptr_cleared;
+    }
+
+    inline void advance_queue_local_wptr(uint32_t num_words) {
+        *this->wptr = wrap_ptr(*this->wptr + num_words);
     }
 
     inline void advance_queue_local_rptr_sent(uint32_t num_words)  {
-        if (this->queue_is_input) {
-            this->local_rptr_sent += num_words;
-        } else {
-        *this->local_rptr_sent_update = num_words << REMOTE_DEST_BUF_WORDS_FREE_INC;
-        }
+        *this->rptr_sent = wrap_ptr(*this->rptr_sent + num_words);
     }
 
     inline void advance_queue_local_rptr_cleared(uint32_t num_words)  {
-
-        if (this->queue_is_input) {
-            this->local_rptr_cleared += num_words;
-        } else {
-        *this->local_rptr_cleared_update = num_words << REMOTE_DEST_BUF_WORDS_FREE_INC;
-        }
+        *this->rptr_cleared = wrap_ptr(*this->rptr_cleared + num_words);
     }
 
     inline uint32_t get_queue_data_num_words_occupied() const {
-        return (this->get_queue_local_wptr() - this->get_queue_local_rptr_cleared()) & this->queue_size_mask;
+        return distance(this->get_queue_local_wptr(), this->get_queue_local_rptr_cleared());
     }
 
     inline uint32_t get_queue_data_num_words_free() const {
-        return this->queue_size_words - this->get_queue_data_num_words_occupied();
+        // Important - One word is reserved to distinguish between full and empty.
+        return this->queue_size_words - this->get_queue_data_num_words_occupied() - 1;
```

Comment:
Not a fan of burning a word here to indicate buffer full, but it probably does not hurt perf, and makes pointer management a little bit easy.